code/game/objects/structures/holosign.dm

```diff
@@ -8,22 +8,49 @@
 	max_integrity = 1
 	armor = list("melee" = 0, "bullet" = 50, "laser" = 50, "energy" = 50, "bomb" = 0, "bio" = 0, "rad" = 0, "fire" = 20, "acid" = 20)
 	var/obj/item/holosign_creator/projector
+	var/lifespan = 5 MINUTES
+	var/death_time
+	var/infinite = FALSE //for infinite holosigns
+
+	var/countdown_color
+	var/obj/effect/countdown/holosign/countdown
 
 /obj/structure/holosign/New(loc, source_projector)
 	if(source_projector)
 		projector = source_projector
 		projector.signs += src
 	..()
 
-/obj/structure/holosign/Initialize()
+/obj/structure/holosign/Initialize(mapload, source, new_lifespan)
 	. = ..()
 	alpha = 0
 	SSvis_overlays.add_vis_overlay(src, icon, icon_state, ABOVE_MOB_LAYER, plane, dir, add_appearance_flags = RESET_ALPHA) //you see mobs under it, but you hit them like they are above it
 
+	if(!infinite)
+		START_PROCESSING(SSobj, src)
+
+		if(new_lifespan)
+			lifespan = new_lifespan
+		death_time = world.time + lifespan
+		countdown = new(src)
+		if(countdown_color)
+			countdown.color = countdown_color
+		countdown.start()
+
+/obj/structure/holosign/process()
+	if(death_time < (world.time + 60 SECONDS) && death_time > (world.time + 58 SECONDS)) //there has to be a better way to get this to just trigger once
+		countdown.invisibility = 0
+		playsound(src, 'sound/machines/triple_beep.ogg', 50, TRUE)
+
+	if(death_time < world.time)
+		playsound(src, 'sound/effects/empulse.ogg', 50, TRUE)
+		qdel(src)
+
 /obj/structure/holosign/Destroy()
 	if(projector)
 		projector.signs -= src
 		projector = null
+	qdel(countdown)
```

Comment:
should be QDEL_NULL()

Reply:
actually i'd probably add a Destroy() proc for holosigns here that looks like
/obj/effect/countdown/holosign/Destroy(...)
  if(attached_to)
    var/obj/structure/holosign/H = attached_to
    if(H.countdown)
        H.countdown = null
  return ..()
that way holosigns clean up their ref to the countdown on deletion and the countdown cleans up its ref to the holosign regardless of whether it's deleting because the holosign is (it pays to be safe)